datafusion/functions/src/string/concat.rs

```diff
@@ -65,14 +75,223 @@ impl Default for ConcatFunc {
 
 impl ConcatFunc {
     pub fn new() -> Self {
-        use DataType::*;
         Self {
-            signature: Signature::variadic(
-                vec![Utf8View, Utf8, LargeUtf8],
-                Volatility::Immutable,
-            ),
+            signature: Signature::user_defined(Volatility::Immutable),
         }
     }
+
+    fn concat_arrays(&self, args: &[ColumnarValue]) -> Result<ColumnarValue> {
+        if args.is_empty() {
+            return plan_err!("concat requires at least one argument");
+        }
+
+        // Simple case: single row - use fast path
+        let num_rows = args
+            .iter()
+            .find_map(|arg| match arg {
+                ColumnarValue::Array(array) => Some(array.len()),
+                _ => None,
+            })
+            .unwrap_or(1);
+
+        if num_rows == 1 {
+            return self.concat_arrays_single_row(args);
+        }
+
+        // Multi-row case: process more carefully to avoid blocking
+        let arrays: Result<Vec<Arc<dyn Array>>> = args
+            .iter()
+            .map(|arg| match arg {
+                ColumnarValue::Array(array) => Ok(Arc::clone(array)),
+                ColumnarValue::Scalar(scalar) => scalar.to_array_of_size(num_rows),
+            })
+            .collect();
+
+        let arrays = arrays?;
+
+        // Build result using efficient batched operations
+        self.concat_arrays_multi_row(&arrays, num_rows)
+    }
+
+    /// Fast path for single-row array concatenation
+    fn concat_arrays_single_row(&self, args: &[ColumnarValue]) -> Result<ColumnarValue> {
+        let mut all_elements = Vec::new();
+
+        for arg in args {
+            match arg {
+                ColumnarValue::Array(array) => {
+                    if !array.is_null(0) {
+                        let elements = self.extract_row_elements(array.as_ref(), 0)?;
+                        all_elements.extend(elements);
+                    }
+                }
+                ColumnarValue::Scalar(scalar) => {
+                    let array = scalar.to_array_of_size(1)?;
+                    if !array.is_null(0) {
+                        let elements = self.extract_row_elements(array.as_ref(), 0)?;
+                        all_elements.extend(elements);
+                    }
+                }
+            }
+        }
+
+        if all_elements.is_empty() {
+            return plan_err!("No elements to concatenate");
+        }
```

Comment:
So inputs of `concat([null], [null])` would return an error if I understand this correctly?